src/rmc.rs

```diff
@@ -0,0 +1,426 @@
+/// and the set of nodes about who we know that they have a multisignature
+/// * in the completed phase we only store the multisigned hash.
+enum HashState<'a, H: Signable, MK: MultiKeychain> {
+    SignatureAggregation {
+        signed: Signed<'a, Indexed<H>, MK>,
+        partial: PartiallyMultisigned<'a, H, MK>,
+    },
+    MultisignatureAggregation {
+        multisigned: Multisigned<'a, H, MK>,
+        confirmed: BoolNodeMap,
+    },
+    Completed {
+        _multisigned: Multisigned<'a, H, MK>,
+    },
+}
+
+impl<'a, H: Signable + Clone, MK: MultiKeychain> HashState<'a, H, MK> {
+    fn new(signed: Signed<'a, Indexed<H>, MK>, keychain: &'a MK) -> Self {
+        let partial = signed.clone()._into_partially_multisigned(keychain);
+        HashState::SignatureAggregation { signed, partial }
+    }
+
+    fn handle_signed_hash(
+        maybe_state: Option<Self>,
+        signed_indexed_hash: Signed<'a, Indexed<H>, MK>,
+        node_count: NodeCount,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match maybe_state {
+            Some(HashState::SignatureAggregation {
+                signed,
+                mut partial,
+            }) => {
+                partial.add_signature(signed_indexed_hash);
+                match partial._try_into_complete(keychain) {
+                    Ok(multisigned) => {
+                        multisigned_hashes_tx
+                            .unbounded_send(multisigned.clone())
+                            .expect("should succeed");
+                        let mut confirmed = BoolNodeMap::with_capacity(node_count);
+                        confirmed.set(keychain.index());
+                        Some(HashState::MultisignatureAggregation {
+                            multisigned,
+                            confirmed,
+                        })
+                    }
+                    Err(e) => Some(HashState::SignatureAggregation {
+                        signed,
+                        partial: e.partial,
+                    }),
+                }
+            }
+            _ => maybe_state,
+        }
+    }
+
+    fn handle_multisigned_hash(
+        maybe_state: Option<Self>,
+        multisigned_hash: Multisigned<'a, H, MK>,
+        signer_index: NodeIndex,
+        node_count: NodeCount,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match maybe_state {
+            Some(HashState::Completed { .. }) => maybe_state,
+            Some(HashState::MultisignatureAggregation {
+                multisigned,
+                mut confirmed,
+                ..
+            }) => {
+                confirmed.set(signer_index);
+                if 3 * confirmed.true_indices().count() > 2 * node_count.0 {
+                    Some(HashState::Completed {
+                        _multisigned: multisigned,
```

Comment:
Why this underscore?

Reply:
At that moment we didn't use this field yet. I have added `pub fn get_multisigned(&self, hash: &H) -> Option<Multisigned<'a, H, MK>>`, so this will be used now.